src/Sylius/Component/Core/Taxation/OrderShipmentByZoneTaxesApplicator.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Taxation;
+
+use Sylius\Component\Addressing\Model\ZoneInterface;
+use Sylius\Component\Core\Model\AdjustmentInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Order\Factory\AdjustmentFactoryInterface;
+use Sylius\Component\Taxation\Calculator\CalculatorInterface;
+use Sylius\Component\Taxation\Resolver\TaxRateResolverInterface;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class OrderShipmentByZoneTaxesApplicator implements OrderTaxesByZoneApplicatorInterface
+{
+    /**
+     * @var CalculatorInterface
+     */
+    private $calculator;
+
+    /**
+     * @var AdjustmentFactoryInterface
+     */
+    private $adjustmentFactory;
+
+    /**
+     * @var TaxRateResolverInterface
+     */
+    private $taxRateResolver;
+
+    /**
+     * @param CalculatorInterface $calculator
+     * @param AdjustmentFactoryInterface $adjustmentFactory
+     * @param TaxRateResolverInterface $taxRateResolver
+     */
+    public function __construct(CalculatorInterface $calculator, AdjustmentFactoryInterface $adjustmentFactory, TaxRateResolverInterface $taxRateResolver)
+    {
+        $this->calculator = $calculator;
+        $this->adjustmentFactory = $adjustmentFactory;
+        $this->taxRateResolver = $taxRateResolver;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function apply(OrderInterface $order, ZoneInterface $zone)
+    {
+        $lastShipment = $order->getLastShipment();
+        if (!$lastShipment) {
+            return;
+        }
+
+        $taxRate = $this->taxRateResolver->resolve($lastShipment->getMethod(), array('zone' => $zone));
```

Comment:
Let's move this after check if there are any shipping adjustments (it's more costly method).